### ISSUE TITLE
Feature request: Use monte_carlo_csiszar_f_divergence for more than one variable

### ISSUE THREAD

**Issue by tillahoffmann** [tensorflow/probability#223]
Using [`monte_carlo_csiszar_f_divergence`](https://www.tensorflow.org/probability/api_docs/python/tfp/vi/monte_carlo_csiszar_f_divergence) to perform inference for a single distribution works as expected (see example below). 

However, I couldn't figure how to make use of `monte_carlo_csiszar_f_divergence` for more complex models that involves different distributions (e.g. one normal, one gamma). Any advice would be much appreciated. #147 may be related?

```python
import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import distributions as tfd
import functools as ft
import numpy as np


np.random.seed(1)
samples = np.random.normal(-2, 1, 5)  # Generate some "data"


def evaluate_log_joint(samples, mean, *extra_variables):
    # QUESTION: How can I make use of extra_variables when evaluating the log joint?
    rv_mean = tfd.Normal(0, 1)  # Prior for the mean
    rv_samples = tfd.Normal(mean, 1)  # Likelihood
    return tf.reduce_mean(  # Take the mean over monte carlo samples
        tf.reduce_sum(rv_samples.log_prob(samples), axis=-1) + 
        rv_mean.log_prob(mean)
    )


with tf.Graph().as_default() as graph:
    surrogate_posterior = tfd.Normal(
        tf.Variable([0.]),
        tf.Variable([1.])
    )

    elbo_loss = tfp.vi.monte_carlo_csiszar_f_divergence(
        f=tfp.vi.kl_reverse,
        p_log_prob=ft.partial(evaluate_log_joint, samples),
        q=surrogate_posterior,  # QUESTION: How can I pass in multiple nodes in a standard factor graph?
        num_draws=1000,
        name='elbo_loss'
    )
    
    train = tf.train.AdamOptimizer(.1).minimize(elbo_loss)
    init_op = tf.global_variables_initializer()
    
    
sess = tf.Session(graph=graph)
sess.run(init_op)

losses = []

for i in range(200):
    _, loss = sess.run([train, elbo_loss])
    losses.append(loss)
    
    
# Calculate the posterior parameters exactly
neff = samples.size + 1
loc = np.sum(samples) / neff
scale = 1 / np.sqrt(neff)

print(loc, sess.run(surrogate_posterior.loc))  # -1.6205239658469637 [-1.6224545]
print(scale, sess.run(surrogate_posterior.scale))  # 0.4082482904638631 [0.40540358]
```

**Comment by axch** [comment]
Good one.  Doesn't look like there's any way to use the software as it stands to do that.  The best I have for you right now is that [monte_carlo_csiszar_f_divergence](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/vi/csiszar_divergence.py#L899-L903) is not a very complex function, so you could write your own variant.

For the future, I will edit this issue into a feature request to add the capability.  If you feel like submitting a PR, we would welcome it!

**Comment by tillahoffmann** [comment]
@axch, thanks for the update. Do you have recommendations on how to perform variational inference using tensorflow probability for multivariate models?

**Comment by csuter** [comment]
I'm gonna remove the 'good first issue' tag, because I'm fairly sure solving this completely will require some thoughtful and non-trivial API design discussion. Namely, it's a question of defining and aligning the true and variational model, when the structure and alignment are not as simple as, say, mean-field Gaussian.

**Comment by davmre** [comment]
This was addressed in 499827efa11b55f44fa0d5ef0432f3e1eebeff01 (and several CLs leading up to it). It's now possible to pass JointDistribution objects to `tfp.vi.monte_carlo_variational_loss`, which is a renamed and updated version of `monte_carlo_csiszar_f_divergence`; the updated code will be included in the upcoming TFP 0.8 release.

Closing this issue; feel free to reopen if needed.
